Port triangular_solve to structured kernel

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66401
* #62180
* #65606
* #64391
* #61536
* #63948
* #61858
* __->__ #61857

A few updates to internal code that allow marking triangular_solve as structured.

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233

Differential Revision: [D31928687](https://our.internmc.facebook.com/intern/diff/D31928687)

IvanYashchuk committed:
Port triangular_solve to structured kernel

A few updates to internal code that allow to mark triangular_solve as structured.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow to mark triangular_solve as structured.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow to mark triangular_solve as structured.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow to mark triangular_solve as structured.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow to mark triangular_solve as structured.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow to mark triangular_solve as structured.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow to mark triangular_solve as structured.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow to mark triangular_solve as structured.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow to mark triangular_solve as structured.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow to mark triangular_solve as structured.

[ghstack-poisoned]

IvanYashchuk commented:
There is going to be a conflict with https://github.com/pytorch/pytorch/pull/63567. That PR should be merged first.

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow to mark triangular_solve as structured.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow to mark triangular_solve as structured.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow to mark triangular_solve as structured.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow to mark triangular_solve as structured.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow to mark triangular_solve as structured.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow to mark triangular_solve as structured.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow to mark triangular_solve as structured.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow to mark triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233 @Lezcano

[ghstack-poisoned]

IvanYashchuk commented:
@ysiraichi, could you please take a look at this PR? This is my first attempt at porting to a function to a structured kernel.

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow to mark triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233 @Lezcano

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow to mark triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233 @Lezcano

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow to mark triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233 @Lezcano

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow marking triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow marking triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow marking triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow marking triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow marking triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233

[ghstack-poisoned]

ysiraichi commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1632,60 +1657,38 @@ static std::tuple<Tensor&, Tensor&> triangular_solve_out_info(
-  triangular_solve_out_info(result, clone_A, infos, A_broadcasted, self_broadcasted, upper, transpose, unitriangular);
+  if (copy_needed) {
+    Tensor result_tmp = at::empty({0}, self.options());
+    Tensor clone_A_tmp = at::empty({0}, self.options());
```
Shouldn't it be `A.options()` (not sure if it makes any difference, though)?

ysiraichi commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1593,31 +1624,25 @@ The result of the computation is saved in-place in 'result' tensor,
-  TORCH_INTERNAL_ASSERT(input.device() == clone_input.device());
-  TORCH_INTERNAL_ASSERT(input.device() == infos.device());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(input.dim() >= 2);
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(input.size(-2) == input.size(-1));
```
Are we doing this check in the META function?

ysiraichi commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1593,31 +1624,25 @@ The result of the computation is saved in-place in 'result' tensor,
-  TORCH_INTERNAL_ASSERT(input.device() == result.device());
-  TORCH_INTERNAL_ASSERT(input.device() == clone_input.device());
-  TORCH_INTERNAL_ASSERT(input.device() == infos.device());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(input.dim() >= 2);
```
Nit: do we need these checks? I mean, is this function called by any others?

ysiraichi commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -203,6 +204,36 @@ extern "C" void strsm_(char *side, char *uplo, char *trans, char *diag, int *n,
+           "torch.triangular_solve: Expected A to have at least 2 dimensions, but it has ", A.dim(), " dimensions instead");
+
+  Tensor self_broadcasted, A_broadcasted;
+  std::tie(self_broadcasted, A_broadcasted) = at::native::_linalg_broadcast_batch_dims(self, A, "triangular_solve");
```
This function could benefit from precomputed `Tensor` values (we wouldn't have to call `_linalg_broadcast_batch_dims` two times. But, AFAIU, they aren't yet supported, right @ezyang ?

ezyang replied:
Precomputed values didn't land, but Tensors should never be precomputed!

ezyang replied:
You only make use of sizes() from the broadcasted elements. `_linalg_broadcast_batch_dims` should get refactored to have a variant that only returns sizes.

ysiraichi commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1593,31 +1624,25 @@ The result of the computation is saved in-place in 'result' tensor,
-  TORCH_INTERNAL_ASSERT(infos.scalar_type() == at::kInt);
-  TORCH_INTERNAL_ASSERT(infos.numel() == std::max<int64_t>(1, batchCount(input)));
-  TORCH_INTERNAL_ASSERT(infos.is_contiguous());
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(input.scalar_type() == other.scalar_type());
```
Are we doing this check in the META function?

ysiraichi reviewed (commented):
Hi @IvanYashchuk. Really nice PR! I've left a few minor comments. Let me know if something is unclear.

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow marking triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233 @Lezcano

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow marking triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233 @Lezcano

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow marking triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow marking triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow marking triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow marking triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow marking triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow marking triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow marking triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow marking triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow marking triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr @IvanYashchuk xwang233

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow marking triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow marking triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow marking triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow marking triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow marking triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233

[ghstack-poisoned]

bdhirsh reviewed (approved):
LGTM!

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow marking triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow marking triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow marking triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow marking triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Port triangular_solve to structured kernel"


A few updates to internal code that allow marking triangular_solve as structured.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233

[ghstack-poisoned]
